README.md
---

## uxcore-popover [![Dependency Status](http://img.shields.io/david/uxcore/uxcore-popover.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-popover) [![devDependency Status](http://img.shields.io/david/dev/uxcore/uxcore-popover.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-popover#info=devDependencies) 

## TL;DR

uxcore-popover ui component for react

#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-popover
$ cd uxcore-popover
$ npm install
$ gulp server
```

## Usage
```javascript
let overlay = <div>
                <div className="demoContent">
                    <i className="kuma-icon kuma-icon-information"></i> 
                    <span>你确认要删除此项数据吗？</span>
                </div>
                <div className="demoButton">
                    <Button size="medium">确定</Button>
                    <Button size="medium" color="orange">取消</Button>
                </div>
              </div>;
return (
    <div>
        <h2>Demo 1. 悬浮弹出</h2>
        <div style={{marginLeft: 60}}>
            <Popover placement="topLeft" title={'上左'} overlay={overlay} onOk={this.test}>
            <Button>上左</Button>
            </Popover>
            <Popover placement="top" title={'上边'} overlay={overlay}>
            <Button>上边</Button>
            </Popover>
            <Popover placement="topRight" title={'上右'} overlay={overlay}>
            <Button>上右</Button>
            </Popover>
        </div>
        <div style={{width: 60, float: 'left'}}>
            <Popover placement="leftTop" title={'左上'} overlay={overlay}>
            <Button>左上</Button>
            </Popover>
            <Popover placement="left" title={'左边'} overlay={overlay}>
            <Button>左边</Button>
            </Popover>
            <Popover placement="leftBottom" title={'左下'} overlay={overlay}>
            <Button>左下</Button>
            </Popover>
        </div>
        <div style={{width: 60, marginLeft: 270}}>
            <Popover placement="rightTop" title={'右上'} overlay={overlay}>
            <Button>右上</Button>
            </Popover>
            <Popover placement="right" title={'右边'} overlay={overlay}>
            <Button>右边</Button>
            </Popover>
            <Popover placement="rightBottom" title={'右下'} overlay={overlay}>
            <Button>右下</Button>
            </Popover>
        </div>
        <div style={{marginLeft: 60, clear: 'both'}}>
            <Popover placement="bottomLeft" title={'下左'} overlay={overlay}>
            <Button>下左</Button>
            </Popover>
            <Popover placement="bottom" title={'下边'} overlay={overlay}>
            <Button>下边</Button>
            </Popover>
            <Popover placement="bottomRight" title={'下右'} overlay={overlay}>
            <Button>下右</Button>
            </Popover>
        </div>
        <h2>Demo 2. 点击弹出</h2>
        <Popover title="测试" overlay={overlay} placement="right" trigger="click">
            <Button>从右弹出</Button>
        </Popover>
        <h2>Demo 3. 没有标题</h2>
        <Popover overlay={overlay} placement="right">
            <Button>从右弹出</Button>
        </Popover>
        <h2>Demo 4. 监听弹窗展开/收起状态</h2>
        <Popover overlay={overlay} placement="bottom" onVisibleChange={me.onVisibleChange.bind(me)}>
             <Button>从下弹出</Button>
        </Popover>
        <h2>Demo 5. 手动控制显示状态</h2>
        <Popover overlay={overlay} placement="bottom" visible={me.state.visible}>
            <Button onMouseOver={me.onChange.bind(me, true)} onMouseLeave={me.onChange.bind(me, false)}>手动控制</Button>
        </Popover>
    </div>
```

### demo
http://uxcore.github.io/uxcore/components/uxcore-popover/

## API

## Props

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|overlay|string|required|-|弹窗内容|
|overlayClassName|string|optional|-|弹窗的额外类名|
|visible|boolean|optional|-|是否展开弹窗内容，如果不填则自动处理；填写则需要手动处理|
|title|string|optional|-|弹窗标题，不传入则没有标题栏|
|placement|string|optional|top|弹窗从那个方向弹出|
|prefixCls|string|optional|kuma-popup|类名前缀，用于定制化|
|trigger|string|optional|hover|触发弹窗的方式，有 hover 和 click 两种|
|delay|number|optional|0|hover延迟显示时间，单位ms|
|showButton|boolean|optional|false|是否显示内置按钮|
|onOk|function(hideCallback)|optional|noop|内置的确定按钮的回调，当想要关闭 popover 时需调用 hideCallback|
|onCancel|function|optional|noop|内置的取消按钮的回调|
|okText|string|optional|"确定"|内置的确定按钮显示的文案|
|cancelText|string|optional|"取消"|内置的取消按钮显示的文案|
|onVisibleChange|function(isDisplay)|optional|noop|显示状态发生改变的回调，isDisplay表示弹窗是否展开|
|getTooltipContainer|function|optional| - | 返回一个 dom 节点，作为 tooltip 渲染的容器 |




HISTORY.md
# HISTORY

## 0.6.3

* `FIXED` button prefixCls

## 0.5.5

* `CHANGED` update `uxcore-button` to `^0.4.0`


## 0.5.4

* `CHANGED` arrow position imporvement
* `CHANGED` update `rc-tooltip` to `^3.0.0`

## 0.5.3

* `CHANGED` react to 15.x
* `FIXED` JSHint issues
* `NEW` tests

## 0.5.2

* `CHANGED` support new prop `align`

## 0.5.1

* `CHANGED` add new prop `getTooltipContainer`

## 0.5.0

* `CHANGED` new style

## 0.4.3

* `CHANGED` animation direction change
* `CHANGED` use global transition ease

## 0.4.0

* `CHANGED` new animation

## 0.3.5

* `FIXED` prop `delay` fail to work
* `CHANGED` new scaffold
* `CHANGED` airbnb javascript style lint

## 0.3.4

`NEW` add new prop `visible`

## 0.3.2

`NEW` add default transition name

## 0.3.0

`CHANGED` open transitionName config

## 0.2.3

`CHANGED` update scaffold

## 0.2.2
`CHANGED` add react require to be compatible with nodejs

## 0.2.1
`CHANGED` add more direction support

## 0.2.0

`CHANGED` add hide callback in `onOK`

## 0.1.2

`NEW` add overlayClassName props

## 0.1.1

`CHANGED` update rc-tooltip denpendency to `~3.0.0`
`FIXED` #3



tests/index.js
/**
 * only require other specs here
 */

import spec from './Popover.spec';


tests/Popover.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils, { Simulate } from 'react-dom/test-utils';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Popover from '../src';

describe('Popover', () => {

  it('Could be rendered without overlay and with visiable true', function () {
    const component = TestUtils.renderIntoDocument(
      <Popover
        placement="right"
        title={'右边'}
        prefixCls="unit-test"
        showButton={true}
        visible={true}
      >
        <span className="overlay-inner-text">右边</span>
      </Popover>
    );
    expect(component.state.visible).equal(true);
  });

  it('Could be rendered with visiable false', function () {
    let onCancleCalled = false;
    let onVisibleChangeCalled = false;
    let onOkCalled = false;
    const props = {
      placement: 'right',
      title: '右边',
      prefixCls: 'unit-test',
      showButton: true,
      visible: false,
      onCancle: () => {onCancleCalled = true},
      onVisibleChange: () => {onVisibleChangeCalled = true},
      onOk: (cb) => {
        cb();
        onOkCalled = true;
      }
    };
    const component = TestUtils.renderIntoDocument(
      <Popover
        {...props}
      >
        <span className="overlay-inner-text">右边</span>
      </Popover>
    );
    expect(component.state.visible).equal(false);
    component.setState({ ...component.state, visible: true }, () => {
      setTimeout(() => {
        expect(component.state.visible).equal(true);
      }, 0)
      component.handleOkClick();
      component.handleCancelClick();
      setTimeout(() => {
        expect(onVisibleChangeCalled).equal(true);
        expect(onCancleCalled).equal(true);
        expect(onOkCalled).equal(true);
      }, 0);
      component.constructor.getDerivedStateFromProps({...props, visible: true}, {})
    });
  });

  it('Could be rendered with overlay without visiable', function () {
    let onCancleCalled = false;
    let onVisibleChangeCalled = false;
    let onOkCalled = false;
    const overlay = (<div>
      <div className="demoContent">
        <span>目标删除后将不可恢复，如有子目标将会删除！</span>
      </div>
    </div>);
    const props = {
      placement: "right",
      title: '右边',
      overlay: overlay,
      onCancle:  () => { onCancleCalled = true },
      onVisibleChange: () => { onVisibleChangeCalled = true },
      onOk: (cb) => {
        cb();
        onOkCalled = true;
      }
    }
    const component = TestUtils.renderIntoDocument(
      <Popover
        {...props}
      >
        <span>右边</span>
      </Popover>
    );
    expect(component.state.visible).equal(false);
    component.constructor.getDerivedStateFromProps({...props, visible: true}, {});
    component.handleOkClick();
    component.handleCancelClick();
    component.handleVisibleChange();
    setTimeout(() => {
      expect(onVisibleChangeCalled).equal(true);
      expect(onCancleCalled).equal(true);
      expect(onOkCalled).equal(true);
    }, 200);
  });

});

src/Popover.js
/**
 * Popover Component for uxcore
 * @author eternalsky
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

let React = require('react');
let ReactDOM = require('react-dom');

let Tooltip = require("rc-tooltip");
let Button = require("uxcore-button");

class Popover extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            visible: 'visible' in props ? props.visible : false
        }
    }

    componentWillReceiveProps(nextProps) {
        let me = this;

        if (('visible' in nextProps) && (nextProps.visible !== me.props.visible)) {
            me.setState({
                visible: nextProps.visible
            });
        }
        
    }

    handleOkClick() {
        let me = this;
        me.props.onOk(() => {
            if (!('visible' in me.props)) {
                me.setState({
                    visible: false
                })
            }
        })
    }

    handleCancelClick() {
        let me = this;

        if (!('visible' in me.props)) {
            me.setState({
                visible: false
            }, () => {
                me.props.onCancel()
            });
        } else {
            me.props.onCancel()
        }
    }

    handleVisibleChange(visible) {
        let me = this;
        if (!('visible' in me.props)) {
            me.setState({
                visible: visible
            }, () => {
                me.props.onVisibleChange(visible);
            })
        } else {
            me.props.onVisibleChange(visible);
        }
    }

    _renderButton() {
        let me = this;
        if (me.props.showButton) {
            return <div className={me.props.prefixCls + "-button-group"}>
                        <Button size="small" type="primary" onClick={me.handleOkClick.bind(me)}>{me.props.okText}</Button>
                        <Button size="small" type="secondary" onClick={me.handleCancelClick.bind(me)}>{me.props.cancelText}</Button>
                   </div>
        }
    }

    _generateOverlay() {
        let me = this;
        let arr = [];
        if (me.props.title) {
            arr.push(<div key="title" className={me.props.prefixCls + '-title'}>
                        {me.props.title}
                    </div>);
        }
        if (me.props.overlay) {
            arr.push(<div key="content" className={me.props.prefixCls + '-content'}>
                        {me.props.overlay}
                        {me._renderButton()}
                    </div>);
        }
        return <div>
                {arr}
               </div>
    }

    render() {
        let me = this;

        return (
            <Tooltip placement={me.props.placement}
                   arrowContent={me.props.arrowContent}
                   overlayClassName={me.props.overlayClassName}
                   prefixCls={me.props.prefixCls}
                   visible={me.state.visible}
                   onVisibleChange={me.handleVisibleChange.bind(me)}
                   mouseEnterDelay={me.props.delay}
                   transitionName={me.props.transitionName}
                   trigger={me.props.trigger}
                   overlay={me._generateOverlay()}>
                {me.props.children}
            </Tooltip>
        );
    }
}

Popover.defaultProps = {
    prefixCls: 'kuma-popover',
    transitionName: 'flip',
    delay: 0,
    placement: "top",
    trigger: "hover",
    onOk: (cb) => {cb()},
    onCancel: () => {},
    okText: "确定",
    cancelText: "取消",
    showButton: false,
    arrowContent: <div className="kuma-popover-arrow-inner"></div>,
    onVisibleChange: () => {}
}

// http://facebook.github.io/react/docs/reusable-components.html
Popover.propTypes = {
    prefixCls: React.PropTypes.string,
    delay: React.PropTypes.number,
    overlayClassName: React.PropTypes.string,
    placement: React.PropTypes.oneOf(['top', 'bottom', 'left', 'right', 'topLeft', 'bottomLeft', 'leftTop', 'rightTop', 'topRight', 'bottomRight', 'leftBottom', 'rightBottom']),
    trigger: React.PropTypes.oneOf(['hover', 'click']),
    onOk: React.PropTypes.func,
    onCancel: React.PropTypes.func,
    okText: React.PropTypes.string,
    cancelText: React.PropTypes.string,
    showButton: React.PropTypes.bool,
    onVisibleChange: React.PropTypes.func
}


Popover.displayName = "Popover";

module.exports = Popover;


webpack.dev.js
var fs = require('fs');
var webpack = require('webpack');

// 扫描uxcore组件目录下的所有module
function getUxcoreModuleAlias() {
    var alias = {};

    // 判断是否存在uxcore目录
    if (!fs.existsSync('./uxcore')) return alias;

    var modules = fs.readdirSync('./uxcore');
    modules.forEach(function (name) {
        alias[name] = [process.cwd(), 'uxcore', name, 'src'].join('/');
    });
    return alias;
}

module.exports = {
    cache: false,
    entry: {
        demo: './demo/index'
    },
    output: {
        path: './dist',
        filename: "[name].js",
        sourceMapFilename: "[name].js.map"
    },
    devtool: '#source-map', // 这个配置要和output.sourceMapFilename一起使用
    module: {
        loaders: [
            {

                test: /\.js(x)*$/,
                // uxcore以外的modules都不需要经过babel解析
                exclude: function (path) {
                    var isNpmModule = !!path.match(/node_modules/);
                    var isUxcore = !!path.match(/node_modules\/uxcore/);
                    return isNpmModule & !isUxcore;
                },
                loader: 'babel-loader?stage=1'
            }
        ]
    },
    resolve: {
       // alias: getUxcoreModuleAlias()
    },
    externals: {
        react: 'var React', // 相当于把全局的React作为模块的返回 module.exports = React;
        "react-dom": 'var ReactDOM'
    },
    plugins: [
        new webpack.DefinePlugin({
          __LOCAL__: true, // 本地环境
          __DEV__:   true, // 日常环境
          __PRO__:   false // 生产环境
        })
    ]
};

demo/PopoverDemo.js
/**
 * Popover Component Demo for uxcore
 * @author eternalsky
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

let classnames = require('classnames');
let Button = require('uxcore-button');

let Popover = require('../src');

class Demo extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            visible: false
        }
    }

    test() {
        alert('test');
    }
    onOk( cb ){
        cb()
    }

    onChange(visible) {
        this.setState({
            visible: visible
        })
    }

    onVisibleChange(visible) {
        if (!visible) {
            this.setState({
                visible: visible
            })
        }
        console.log(visible)
    }

    render() {
        let me = this;
        let overlay = <div>
                        <div className="demoContent">
                            <i className="kuma-icon kuma-icon-information"></i> 
                            <span>你确认要删除此项数据吗？</span>
                        </div>
                      </div>;
        return (
            <div>
                <h2>Demo 1. 悬浮弹出</h2>
                <div style={{marginLeft: 100}}>
                    <div style={{marginLeft: 60}}>
                        <Popover placement="topLeft" title={'上左'} overlay={overlay} onOk={this.test}>
                        <Button>上左</Button>
                        </Popover>
                        <Popover placement="top" title={'上边'} overlay={overlay}>
                        <Button>上边</Button>
                        </Popover>
                        <Popover placement="topRight" title={'上右'} overlay={overlay}>
                        <Button>上右</Button>
                        </Popover>
                    </div>
                    <div style={{width: 60, float: 'left'}}>
                        <Popover placement="leftTop" title={'左上'} overlay={overlay}>
                        <Button>左上</Button>
                        </Popover>
                        <Popover placement="left" title={'左边'} overlay={overlay}>
                        <Button>左边</Button>
                        </Popover>
                        <Popover placement="leftBottom" title={'左下'} overlay={overlay}>
                        <Button>左下</Button>
                        </Popover>
                    </div>
                    <div style={{width: 60, marginLeft: 270}}>
                        <Popover placement="rightTop" title={'右上'} overlay={overlay}>
                        <Button>右上</Button>
                        </Popover>
                        <Popover placement="right" title={'右边'} overlay={overlay}>
                        <Button>右边</Button>
                        </Popover>
                        <Popover placement="rightBottom" title={'右下'} overlay={overlay}>
                        <Button>右下</Button>
                        </Popover>
                    </div>
                    <div style={{marginLeft: 60, clear: 'both'}}>
                        <Popover placement="bottomLeft" title={'下左'} overlay={overlay}>
                        <Button>下左</Button>
                        </Popover>
                        <Popover placement="bottom" title={'下边'} overlay={overlay}>
                        <Button>下边</Button>
                        </Popover>
                        <Popover placement="bottomRight" title={'下右'} overlay={overlay}>
                        <Button>下右</Button>
                        </Popover>
                    </div>
                </div>
                <h2>Demo 2. 点击弹出 & 内置按钮</h2>
                <Popover title="测试" overlay={overlay} placement="right" trigger="click" onOk={me.onOk} showButton={true}>
                    <Button>从右弹出</Button>
                </Popover>
                <h2>Demo 3. 没有标题</h2>
                <Popover overlay={overlay} placement="right">
                    <Button>从右弹出</Button>
                </Popover>
                <h2>Demo 4. 监听弹窗展开/收起状态</h2>
                <Popover overlay={overlay} placement="bottom" onVisibleChange={me.onVisibleChange.bind(me)}>
                    <Button>从下弹出</Button>
                </Popover>
                <h2>Demo 5. 手动控制显示状态</h2>
                <Popover overlay={overlay} placement="bottom" visible={me.state.visible} onVisibleChange={me.onVisibleChange.bind(me)}>
                    <Button onMouseOver={me.onChange.bind(me, true)}>手动控制</Button>
                </Popover>
            </div>
        );
    }
};

module.exports = Demo;
